src/protocol/vm/state.rs

```diff
@@ -139,52 +135,63 @@ impl VMPoolState<PreCachedDB> {
+            for token_address in &self.tokens {
+                let info = AccountInfo {
+                    balance: Default::default(),
+                    nonce: 0,
+                    code_hash: KECCAK_EMPTY,
+                    code: Some(mocked_contract_bytecode.clone()),
+                };
+                engine.state.init_account(
+                    Address::parse_checksummed(to_checksum(token_address, None), None).map_err(
+                        |_| {
+                            SimulationError::EncodingError(
+                                "Checksum for token address must be valid".into(),
+                            )
+                        },
+                    )?,
+                    info,
+                    None,
+                    false,
+                );
+            }
+
             engine.state.init_account(
-                "0x0000000000000000000000000000000000000000"
-                    .parse()
-                    .unwrap(),
+                *EXTERNAL_ACCOUNT,
                 AccountInfo {
-                    balance: Default::default(),
+                    balance: *MAX_BALANCE,
                     nonce: 0,
                     code_hash: KECCAK_EMPTY,
                     code: None,
                 },
                 None,
                 false,
             );
+
+            let zero_account_info = AccountInfo {
+                balance: Default::default(),
+                nonce: 0,
+                code_hash: KECCAK_EMPTY,
+                code: None,
+            };
+
             engine.state.init_account(
-                rAddress::parse_checksummed("0x0000000000000000000000000000000000000004", None)
+                rAddress::parse_checksummed("0x0000000000000000000000000000000000000000", None)
                     .expect("Invalid checksum for external account address"),
-                AccountInfo {
-                    balance: Default::default(),
-                    nonce: 0,
-                    code_hash: KECCAK_EMPTY,
-                    code: None,
-                },
+                zero_account_info.clone(),
                 None,
                 false,
             );
-            let adapter_contract_code =
-                get_contract_bytecode(&adapter_contract_path).map_err(SimulationError::AbiError)?;
-
             engine.state.init_account(
-                rAddress::parse_checksummed(ADAPTER_ADDRESS.to_string(), None)
+                rAddress::parse_checksummed("0x0000000000000000000000000000000000000004", None)
                     .expect("Invalid checksum for external account address"),
-                AccountInfo {
-                    balance: *MAX_BALANCE,
-                    nonce: 0,
-                    code_hash: B256::from(keccak256(adapter_contract_code.clone().bytes())),
-                    code: Some(adapter_contract_code),
-                },
+                zero_account_info.clone(),
```

Comment:
Maybe we can do this in the engine creation logic? These accounts (0 and 4) are created to enable precompiles in the vm. I think we can create them by default in the engine creation?

Reply:
Hmmm these `account_inits` were originally scattered between the two so we thought it would be better organization to keep them in one place. It's a good argument though - we could instead divide them between the tokens/stateless accounts (which are dynamically determined) in `set_engine`, and the 0 and 4 which should never change and are necessary for precompiles in `create_engine`... if you don't think that's too messy (I'll experiment now)